### PR TITLE
[Bug fix] Try to fix the bug in creating heterogenous graph

### PR DESCRIPTION
**Pull request by caojy1998**
## Description

https://github.com/dmlc/dgl/issues/6087

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Review by Rhett-Ying — CHANGES_REQUESTED** [review]
please add dedicated testcase for the bug

File: python/dgl/core.py [hunk 1]
```diff
@@ -231,7 +231,10 @@ def data_dict_to_list(graph, data_dict, func, target):
             for srctype, _, dsttype in graph.canonical_etypes:
                 if target == "u":
                     src_id = graph.get_ntype_id(srctype)
-                    output_list[src_id] = data_dict[srctype]
+                    if isinstance(data_dict, dict):
+                        output_list[src_id] = data_dict[srctype]
+                    else:
```
Comment (Rhett-Ying) [inline python/dgl/core.py:236]: what's the type of `data_dict` in `else`? tensor?

Comment (caojy1998) [reply python/dgl/core.py:236]: > what's the type of `data_dict` in `else`? tensor?

yes, it's a tensor.

Comment (caojy1998) [reply python/dgl/core.py:236]: > what about other branches like `target == 'v'` or any other code in `data_dict_to_list` which may hit the same issue here?

For the most outer branch(the else branch for `if isinstance(func, fn.BinaryMessageFunction):`, it already take a similar approach as mine.

File: tests/python/common/test_heterograph-apply-edges.py [hunk 2]
```diff
@@ -260,6 +279,21 @@ def _print_error(a, b):
             _test(lhs, rhs, binary_op)
 
 
+@unittest.skipIf(
+    dgl.backend.backend_name != "pytorch", reason="Only support PyTorch for now"
+)
+@parametrize_idtype
+def test_unibipartite_heterograph_apply_edges(idtype):
```
Comment (frozenbugs) [inline tests/python/common/test_heterograph-apply-edges.py:286]: the original bug mentions that it is related to heterogeneous graphs with same source and destination node, should we have a test case for this corner cases?

File: tests/python/common/test_heterograph-apply-edges.py [hunk 3]
```diff
@@ -260,6 +279,21 @@ def _print_error(a, b):
...
+    hg = create_random_hetero(idtype)
+
+    hg.nodes["n1"].data["h"] = F.randn((hg.num_nodes("n1"), 1))
+    hg.nodes["n2"].data["h"] = F.randn((hg.num_nodes("n2"), 1))
+    hg.nodes["n3"].data["h"] = F.randn((hg.num_nodes("n3"), 1))
+
+    assert type(hg.srcdata["h"]) != dict
```
Comment (Rhett-Ying) [inline tests/python/common/test_heterograph-apply-edges.py:293]: it's `defaultdict`?

File: python/dgl/core.py [hunk 4]
```diff
@@ -228,13 +228,20 @@ def data_dict_to_list(graph, data_dict, func, target):
     if isinstance(func, fn.BinaryMessageFunction):
         if target in ["u", "v"]:
             output_list = [None] * graph._graph.number_of_ntypes()
-            for srctype, _, dsttype in graph.canonical_etypes:
+            if not isinstance(data_dict, dict):
```
Comment (Rhett-Ying) [inline python/dgl/core.py:231]: are we assuming `data_dict` is either `defaultdict` or `dict`?

Comment (frozenbugs) [inline python/dgl/core.py:231]: This is weird, the name is data_dict, in which legal case, it can be a non-dict instance?

Comment (Rhett-Ying) [reply python/dgl/core.py:231]: the name indicates dict-like data structure

Comment (caojy1998) [reply python/dgl/core.py:231]: We assume here data_dict is a dict or a tensor.

File: tests/python/common/test_heterograph-apply-edges.py (hunk 3 above)
Comment (caojy1998) [reply tests/python/common/test_heterograph-apply-edges.py:293]: no，it's a tensor here.

File: tests/python/common/test_heterograph-apply-edges.py (hunk 2 above)
Comment (caojy1998) [reply tests/python/common/test_heterograph-apply-edges.py:286]: Yes, my function create_random_hetero create a heterograph with same source node and same dest node.

File: python/dgl/core.py (hunk 4 above)
Comment (caojy1998) [reply python/dgl/core.py:231]: Yes, when there is only 1 source node, the type of data_dict is not a dict. It is just a feature of the source node(maybe a tensor).

File: tests/python/common/test_heterograph-apply-edges.py (hunk 2 above)
Comment (caojy1998) [reply tests/python/common/test_heterograph-apply-edges.py:286]: I will try to modify the bug description to make it more precise.

File: python/dgl/core.py (hunk 4 above)
Comment (frozenbugs) [reply python/dgl/core.py:231]: If so, please update the python doc of the method and add a comment above this line.

Comment (caojy1998) [reply python/dgl/core.py:231]: Shall we change the variable name here?

File: tests/python/common/test_heterograph-apply-edges.py [hunk 5]
```diff
@@ -10,7 +10,9 @@
 import networkx as nx
 import numpy as np
 import pytest
+import scipy.sparse as spsp
 import scipy.sparse as ssp
```
Comment (frozenbugs) [inline tests/python/common/test_heterograph-apply-edges.py:14]: duplication?

File: tests/python/common/test_heterograph-apply-edges.py (hunk 2 above)
Comment (frozenbugs) [reply tests/python/common/test_heterograph-apply-edges.py:286]: if create_random_hetero has something special in it, then it should be called out in the method name.
e.g. create_random_hetero -> create_random_hetero_include_same_source_node.

File: tests/python/common/test_heterograph-apply-edges.py (hunk 5 above)
Comment (caojy1998) [reply tests/python/common/test_heterograph-apply-edges.py:14]: Yes, I have deleted the duplicate one.

**Review by frozenbugs — APPROVED** [review]

File: tests/python/common/test_heterograph-apply-edges.py (hunk 2 above)
Comment (caojy1998) [reply tests/python/common/test_heterograph-apply-edges.py:285]: OK, I will add comment here.

File: python/dgl/core.py [hunk 6]
```diff
@@ -213,7 +213,8 @@ def data_dict_to_list(graph, data_dict, func, target):
         The input graph.
     data_dict : dict[str, Tensor] or dict[(str, str, str), Tensor]]
```
Comment (Rhett-Ying) [inline python/dgl/core.py:214]: add `or Tensor`?

File: python/dgl/core.py [hunk 7]
```diff
@@ -213,7 +213,8 @@ def data_dict_to_list(graph, data_dict, func, target):
...
         Node or edge data stored in DGLGraph. The key of the dictionary
-        is the node type name or edge type name.
+        is the node type name or edge type name. If there is only 1 source
+        node, data_dict is the value of feature not a dict.
```
Comment (Rhett-Ying) [inline python/dgl/core.py:217]: `is the value of feature` indicates `Tensor`? If yes, explain more clearly in docstring?

File: python/dgl/core.py [hunk 8]
```diff
@@ -228,13 +229,22 @@ def data_dict_to_list(graph, data_dict, func, target):
     if isinstance(func, fn.BinaryMessageFunction):
         if target in ["u", "v"]:
             output_list = [None] * graph._graph.number_of_ntypes()
-            for srctype, _, dsttype in graph.canonical_etypes:
+            # If there is only 1 source node, data_dict should be the value of feature,
+            # such as a tensor.
```
Comment (Rhett-Ying) [inline python/dgl/core.py:233]: `such as a tensor` or `namely, a tensor`? can it be other data types besides `tensor`?

File: tests/python/common/test_heterograph-apply-edges.py [hunk 9]
```diff
@@ -260,6 +278,23 @@ def _print_error(a, b):
             _test(lhs, rhs, binary_op)
 
 
+# Here we test heterograph with only 1 source node because the format of node
```
Comment (Rhett-Ying) [inline tests/python/common/test_heterograph-apply-edges.py:281]: this comment looks not clear to me. `only single source node type`? clearly describe the data structure as well?

File: tests/python/common/test_heterograph-apply-edges.py [hunk 10]
```diff
@@ -47,6 +48,23 @@ def create_test_heterograph(idtype):
     return g
 
 
+def create_random_hetero_with_same_source_node(idtype):
```
Comment (Rhett-Ying) [inline tests/python/common/test_heterograph-apply-edges.py:51]: `with_single_node_type`?

File: tests/python/common/test_heterograph-apply-edges.py [hunk 11]
```diff
@@ -260,6 +278,23 @@ def _print_error(a, b):
...
+)
+@parametrize_idtype
+def test_heterograph_with_same_source_node_apply_edges(idtype):
+    hg = create_random_hetero_with_same_source_node(idtype)
+
+    hg.nodes["n1"].data["h"] = F.randn((hg.num_nodes("n1"), 1))
+    hg.nodes["n2"].data["h"] = F.randn((hg.num_nodes("n2"), 1))
+    hg.nodes["n3"].data["h"] = F.randn((hg.num_nodes("n3"), 1))
+
+    assert type(hg.srcdata["h"]) == F.Tensor
```
Comment (Rhett-Ying) [inline tests/python/common/test_heterograph-apply-edges.py:294]: `F.Tensor` is available? or we could just utilize `F.is_tensor()`

File: python/dgl/core.py [hunk 12]
```diff
@@ -228,13 +229,22 @@ def data_dict_to_list(graph, data_dict, func, target):
...
+            if not isinstance(data_dict, dict):
```
Comment (Rhett-Ying) [inline python/dgl/core.py:234]: or `if F.is_tensor(data_dict)`?

File: python/dgl/core.py (hunk 6 above)
Comment (Rhett-Ying) [reply python/dgl/core.py:214]: and is it better to rename `data_dict` to a more general name?

Comment (caojy1998) [reply python/dgl/core.py:214]: OK, I have added.

File: python/dgl/core.py (hunk 8 above)
Comment (caojy1998) [reply python/dgl/core.py:233]: OK, I have changed it. I should be a tensor with 1 source node.

File: tests/python/common/test_heterograph-apply-edges.py (hunk 10 above)
Comment (caojy1998) [reply tests/python/common/test_heterograph-apply-edges.py:51]: OK, I have change it.

File: tests/python/common/test_heterograph-apply-edges.py (hunk 11 above)
Comment (caojy1998) [reply tests/python/common/test_heterograph-apply-edges.py:294]: OK，I changed to torch.Tensor here.

File: tests/python/common/test_heterograph-apply-edges.py (hunk 9 above)
Comment (caojy1998) [reply tests/python/common/test_heterograph-apply-edges.py:281]: OK, I have modified it.

File: python/dgl/core.py (hunk 7 above)
Comment (caojy1998) [reply python/dgl/core.py:217]: Modified

File: python/dgl/core.py [hunk 13]
```diff
@@ -211,9 +211,10 @@ def data_dict_to_list(graph, data_dict, func, target):
     -------------
     graph :  DGLGraph
         The input graph.
-    data_dict : dict[str, Tensor] or dict[(str, str, str), Tensor]]
+    data_dict : dict[str, Tensor] or dict[(str, str, str), Tensor]] or Tensor
         Node or edge data stored in DGLGraph. The key of the dictionary
-        is the node type name or edge type name.
+        is the node type name or edge type name. If there is only 1 source
+        node, data_dict is the value of feature(a Tensor) not a dict.
```
Comment (Rhett-Ying) [inline python/dgl/core.py:217]: `single node type`

File: python/dgl/core.py [hunk 14]
```diff
@@ -228,13 +229,22 @@ def data_dict_to_list(graph, data_dict, func, target):
     if isinstance(func, fn.BinaryMessageFunction):
         if target in ["u", "v"]:
             output_list = [None] * graph._graph.number_of_ntypes()
-            for srctype, _, dsttype in graph.canonical_etypes:
+            # If there is only 1 source node, data_dict should be the value of feature,
```
Comment (Rhett-Ying) [inline python/dgl/core.py:232]: single source node type

Comment (caojy1998) [reply python/dgl/core.py]: OK, I have changed it.

File: python/dgl/core.py [hunk 15]
```diff
@@ -211,9 +211,10 @@ def data_dict_to_list(graph, data_dict, func, target):
     -------------
     graph :  DGLGraph
         The input graph.
-    data_dict : dict[str, Tensor] or dict[(str, str, str), Tensor]]
+    data_dict : dict[str, Tensor] or dict[(str, str, str), Tensor]] or Tensor
```
Comment (Rhett-Ying) [inline python/dgl/core.py:214]: decided to not rename `data_dict`?

**Review by Rhett-Ying — APPROVED** [review]
